docs/src/locators.md

```diff
@@ -188,121 +129,367 @@ await page.Locator("button").CountAsync();
+# Ensure that "Check me" checkbox is checked.
+page.locator('role=checkbox[checked][name="Check me"]').check()
+```
+
+```java
+// Click the button with the text "submit".
+page.locator("role=button[name=/submit/i]").click();
+
+// Ensure that "Check me" checkbox is checked.
+page.locator("role=checkbox[checked][name=\"Check me\"]").check();
+```
+
+```csharp
+// Click the button with the text "submit".
+await page.Locator("role=button[name=/submit/i]").ClickAsync();
+
+// Ensure that "Check me" checkbox is checked.
+await page.Locator("role=checkbox[checked][name=\"Check me\"]").CheckAsync();
+```
+
+[Learn more about the `role` selector](./selectors.md#role-selector).
+
+### Define explicit contract and use `data-test-id=`
+
+User-facing attributes like text or accessible name can change frequently. In this case it is convenient to define explicit test ids, for example with a `data-test-id` attribute. Playwright has dedicated support for `id`, `data-test-id`, `data-test` and `data-testid` attributes.
+
+```html
+<button data-test-id="directions">Itinéraire</button>
+```
+
+```js
+await page.locator('data-test-id=directions').click();
+```
+
+```java
+page.locator("data-test-id=directions").click();
+```
+
+```python async
+await page.locator('data-test-id=directions').click()
 ```
 
 ```python sync
-# Locate elements, this locator points to a list.
-rows = page.locator("table tr")
+page.locator('data-test-id=directions').click()
+```
 
-# Pattern 1: use locator methods to calculate text on the whole list.
-texts = rows.all_text_contents()
+```csharp
+await page.Locator("data-test-id=directions").ClickAsync();
+```
 
-# Pattern 2: do something with each element in the list.
-count = rows.count()
-for i in range(count):
-  print(rows.nth(i).text_content())
+### Locate by label text
 
-# Pattern 3: resolve locator to elements on page and map them to their text content.
-# Note: the code inside evaluateAll runs in page, you can call any DOM apis there.
-texts = rows.evaluate_all("list => list.map(element => element.textContent)")
+Most form controls usually have dedicated labels that could be conveniently used to interact with the form. Input actions in Playwright automatically distinguish between labels and controls, so you can just locate the label to perform an action on the associated control.
+
+For example, consider the following DOM structure.
+
+```html
+<label for="password">Password:</label><input type="password">
+```
+
+You can target the label with something like `text=Password` and perform the following actions on the password input:
+- `click` will click the label and automatically focus the input field;
+- `fill` will fill the input field;
+- `inputValue` will return the value of the input field;
+- `selectText` will select text in the input field;
+- `setInputFiles` will set files for the input field with `type=file`;
+- `selectOption` will select an option from the select box.
+
+```js
+// Fill the input by targeting the label.
```

Comment:
same here. I suggest to add text outside of code block so we only have to maintain the explanation once